libvirt/tunnel/tunnel.sh

```diff
@@ -1,98 +1,130 @@
-                        -R ${CLUSTER_PORTS["1,API"]}:127.0.0.1:16443  -R ${CLUSTER_PORTS["1,HTTP"]}:127.0.0.1:10080 -R ${CLUSTER_PORTS["1,HTTPS"]}:127.0.0.1:10443 \
-                        -R ${CLUSTER_PORTS["2,API"]}:127.0.0.1:26443  -R ${CLUSTER_PORTS["2,HTTP"]}:127.0.0.1:20080 -R ${CLUSTER_PORTS["2,HTTPS"]}:127.0.0.1:20443 \
-                        -R ${CLUSTER_PORTS["3,API"]}:127.0.0.1:36443  -R ${CLUSTER_PORTS["3,HTTP"]}:127.0.0.1:30080 -R ${CLUSTER_PORTS["3,HTTPS"]}:127.0.0.1:30443 \
-                        -R ${CLUSTER_PORTS["4,API"]}:127.0.0.1:46443  -R ${CLUSTER_PORTS["4,HTTP"]}:127.0.0.1:40080 -R ${CLUSTER_PORTS["4,HTTPS"]}:127.0.0.1:40443 \
-		; then
-			echo "[WARNING] SSH tunnelling failed, retrying..."
+		echo "$(timestamp) [INFO] Setting up a reverse SSH tunnel to connect bastion port "${1:?Bastion service port was not specified}" to VM port "${2:?VM port was not specified}"..."
+		if ! ssh -N -T root@127.0.0.1 -p 2222 -R "$1:127.0.0.1:$2"; then
+			echo "$(timestamp) [WARNING] SSH tunnelling failed, retrying..."
 			sleep 0.1
 		fi
 	done
 }
 
+function pid-exists() {
+	kill -0 $1 2>/dev/null
+	return $?
+}
+
 trap "kill 0" SIGINT
+PID_PORT=-1
+PID_SSH=-1
+
+while true; do
+
+	if [[ ${PID_PORT} > 1 ]] && pid-exists ${PID_PORT}; then
+		echo "$(timestamp) *** [WARNING] Killing old port-forward (${PID_PORT})" >> tunnel.log
+		kill -9 ${PID_PORT}
+	fi
+	if [[ ${PID_SSH} > 1 ]] && pid-exists ${PID_SSH}; then
+		echo "$(timestamp) *** [WARNING] Killing old ssh-tunnel (${PID_SSH})" >> tunnel.log
+		kill -9 ${PID_SSH}
+	fi
 
-# set up port forwarding from the SSH bastion to the local port 2222
-port-forward &
+	# set up port forwarding from the SSH bastion to the local port 2222
+	port-forward 2222 &
+	PID_PORT=$!
 
-# without a better synchonization library, we just need to wait for the port-forward to run
-sleep 5
+	# without a better synchonization library, we just need to wait for the port-forward to run
+	sleep 10s
+
+# ${BASTION_PORTS["LIBVIRT"]}:127.0.0.1:16509 \
+# ${CLUSTER_PORTS["0,API"]}:127.0.0.1:6443   ${CLUSTER_PORTS["0,HTTP"]}:127.0.0.1:80    ${CLUSTER_PORTS["0,HTTPS"]}:127.0.0.1:443 \
+# ${CLUSTER_PORTS["1,API"]}:127.0.0.1:16443  ${CLUSTER_PORTS["1,HTTP"]}:127.0.0.1:10080 ${CLUSTER_PORTS["1,HTTPS"]}:127.0.0.1:10443 \
+# ${CLUSTER_PORTS["2,API"]}:127.0.0.1:26443  ${CLUSTER_PORTS["2,HTTP"]}:127.0.0.1:20080 ${CLUSTER_PORTS["2,HTTPS"]}:127.0.0.1:20443 \
+# ${CLUSTER_PORTS["3,API"]}:127.0.0.1:36443  ${CLUSTER_PORTS["3,HTTP"]}:127.0.0.1:30080 ${CLUSTER_PORTS["3,HTTPS"]}:127.0.0.1:30443 \
+# ${CLUSTER_PORTS["4,API"]}:127.0.0.1:46443  ${CLUSTER_PORTS["4,HTTP"]}:127.0.0.1:40080 ${CLUSTER_PORTS["4,HTTPS"]}:127.0.0.1:40443 \
+
+	# run an SSH tunnel from the port $1 on the SSH bastion (through local port 2222) to local port $1
+	ssh-tunnel $(( $libvirt_port + $CLUSTER_ID )) $libvirt_port &
```

Comment:
For this, let's add a function where we build a big list of (x, y) pairs where x is the BASTION_PORT and y is the LOCAL_PORT